include/nabto/nabto_device_virtual.h

```diff
@@ -0,0 +1,323 @@
+     *
+     * @param device [in] The device context
+     * @return The created virtual connection or NULL on failure
+     */
+    NABTO_DEVICE_DECL_PREFIX NabtoDeviceVirtualConnection* NABTO_DEVICE_API
+        nabto_device_virtual_connection_new(NabtoDevice* device);
+
+    /**
+     * Free a previously allocated virtual connection.
+     * @param connection [in] The connection to free.
+     */
+    NABTO_DEVICE_DECL_PREFIX void NABTO_DEVICE_API
+        nabto_device_virtual_connection_free(NabtoDeviceVirtualConnection* connection);
+
+
+    /**
+     * Close a virtual connection.
+     *
+     * @param connection [in] The connection to close.
+     * @param future [in] Future resolved when the connection is closed
+     */
+    NABTO_DEVICE_DECL_PREFIX void NABTO_DEVICE_API
+        nabto_device_virtual_connection_close(NabtoDeviceVirtualConnection* connection, NabtoDeviceFuture* future);
+
+    /**
+     * Set a device fingerprint on a virtual connection.
+     *
+     * The fingerprint is copied into the virtual connection.
+     *
+     * @param connection [in] The connection to close.
+     * @param fp [in] Fingerprint to set.
+     * @retval NABTO_DEVICE_EC_OK on success
+     */
+    NABTO_DEVICE_DECL_PREFIX NabtoDeviceError NABTO_DEVICE_API
+        nabto_device_virtual_connection_set_device_fingerprint(NabtoDeviceVirtualConnection* connection, const char* fp);
+
+    /**
+     * Set a client fingerprint on a virtual connection.
+     *
+     * The fingerprint is copied into the virtual connection.
+     *
+     * @param connection [in] The connection to close.
+     * @param fp [in] Fingerprint to set.
+     * @retval NABTO_DEVICE_EC_OK on success
+     */
+    NABTO_DEVICE_DECL_PREFIX NabtoDeviceError NABTO_DEVICE_API
+        nabto_device_virtual_connection_set_client_fingerprint(NabtoDeviceVirtualConnection* connection, const char* fp);
+
+    /**
+     * Get the connection reference of a virtual connection.
+     *
+     * @param connection [in]  The virtual connection to get reference for
+     * @return The connection reference of the provided virtual connection.
+     */
+    NABTO_DEVICE_DECL_PREFIX NabtoDeviceConnectionRef  NABTO_DEVICE_API
+        nabto_device_connection_get_connection_ref(NabtoDeviceVirtualConnection* connection);
+
+
+    /**
+     * Test if the connection is virtual.
+     *
+     * @param device [in]  The device
+     * @param ref [in]     The connection reference to query
+     * @return true iff the connection is virtual.
+     */
+    NABTO_DEVICE_DECL_PREFIX bool NABTO_DEVICE_API
+        nabto_device_connection_is_virtual(NabtoDevice* device, NabtoDeviceConnectionRef ref);
+
+
+    typedef struct NabtoDeviceVirtualCoapRequest_ NabtoDeviceVirtualCoapRequest;
+
+    /**
+     * Create a virtual CoAP request.
+     *
+     * @param connection [in]  The virtual connection to make the CoAP request on, the connection needs
+     * to be kept alive until the request has been freed.
```

Comment:
document parameters and lifetime of data. Copy segments such that the caller should not keep these alive.